### PR TITLE
[REF][9.0] currency_rate_update: Refactoring related to PR #318

### PR DESCRIPTION
**Pull request by katyukha**
This is little refactoring based on PR #318  (rebased on current 9.0)

Now if you want to create new "currency getter", you only need to subclass `Currency_getter_interface` class, so new currency getters may be easily implemented in separate addons. Also, currenct currency getters could be easily and safely moved to separate addons if needed.

So there are a few questions:
- should current currency getters be moved to separate addons (or gether them all to only one separate addon?
- or its better to leave them in this base module, but implement new currency getters as separate addons? new addon for each currency getter? or group them by some criteria?

And one note, currency getter "Bank of Canada - noon rates" seems not working, it raised error
`'Error !', 'Exchange data format error for Bank of Canada - EUR !` for me. All other seems to be working, at least no errors were raised.

### CODE REVIEW

**Comment by ghost** [comment]
:+1:  Looks good!

**Comment by max3903** [comment]
@katyukha Can you create an issue and mention it in the README.rst (Known issues section) to explain that the Bank of Canada is not working?

Thanks!

**Comment by fv-clearcorp** [comment]
:+1:

File: currency_rate_update/__openerp__.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
 ##############################################################################
 {
     "name": "Currency Rate Update",
-    "version": "8.0.0.7.0",
+    "version": "9.0.0.7.0",
```
Comment (pedrobaeza) [inline currency_rate_update/__openerp__.py:23]: 9.0.1.0.0

**Comment by rossigee** [comment]
+1 - Tested and works for me. Only issue I had was that initially, when clicking 'Update Now', nothing was happening, despite the logs saying 'Starting to refresh...'.I figured it out in the end, but can I just suggest the attached patch to in the hope it might save others from scratching their heads.

```
diff --git a/currency_rate_update/model/currency_rate_update.py b/currency_rate_update/model/currency_rate_update.py
index 4e7d073..a8fc5a7 100644
--- a/currency_rate_update/model/currency_rate_update.py
+++ b/currency_rate_update/model/currency_rate_update.py
@@ -126,14 +126,17 @@ class CurrencyRateUpdateService(models.Model):
     @api.one
     def refresh_currency(self):
         """Refresh the currencies rates !!for all companies now"""
-        _logger.info(
-            'Starting to refresh currencies with service %s (company: %s)',
-            self.service, self.company_id.name)
         rate_obj = self.env['res.currency.rate']
         company = self.company_id
-        # The multi company currency can be set or no so we handle
-        # The two case
-        if company.auto_currency_up:
+        # The multi company currency can be set or no so we handle both cases
+        if not company.auto_currency_up:
+            _logger.info(
+                'Currency updates disabled (company: %s)',
+                self.company_id.name)
+        else:
+            _logger.info(
+                'Starting to refresh currencies with service %s (company: %s)',
+                self.service, self.company_id.name)
             main_currency = self.company_id.currency_id
             if not main_currency:
                 raise exceptions.Warning(_('There is no main \
```

**Comment by rossigee** [comment]
This also works fine Bitcoin exchange rates using at least the Yahoo service. I've created a 'BTC' currency and added 'BTC' to the array of supported currencies, and this seems to work fine so far. If I remember, I'll submit a separate PR for that later.

**Comment by rossigee** [comment]
FTR, I was able to get a higher degree of accuracy for BTC exchange rates using this patch...

```
diff --git a/currency_rate_update/services/update_service_YAHOO.py b/currency_rate_update/services/update_service_YAHOO.py
index e20319f..5411db4 100644
--- a/currency_rate_update/services/update_service_YAHOO.py
+++ b/currency_rate_update/services/update_service_YAHOO.py
@@ -48,7 +48,12 @@ class YAHOOGetter(CurrencyGetterInterface):
         "SRD", "STD", "SVC", "SYP", "SZL", "THB", "TJS", "TMT", "TND", "TOP",
         "TRY", "TTD", "TWD", "TZS", "UAH", "UGX", "USD", "UYU", "UZS", "VEF",
         "VND", "VUV", "WST", "XAF", "XAG", "XAU", "XCD", "XCP", "XDR", "XOF",
-        "XPD", "XPF", "XPT", "YER", "ZAR", "ZMW", "ZWL"]
+        "XPD", "XPF", "XPT", "YER", "ZAR", "ZMW", "ZWL", "BTC"]
+
+    # We can get a more accurate decimal rate by converting the other way
+    invert_currencies = [
+        "BTC",
+    ]

     def get_updated_currency(self, currency_array, main_currency,
                              max_delta_days):
@@ -60,8 +65,12 @@ class YAHOOGetter(CurrencyGetterInterface):
             currency_array.remove(main_currency)
         for curr in currency_array:
             self.validate_cur(curr)
-            res = self.get_url(url % (main_currency + curr))
-            val = res.split(',')[1]
+            if curr in self.invert_currencies:
+                res = self.get_url(url % (curr + main_currency))
+                val = str(1.0 / float(res.split(',')[1]))
+            else:
+                res = self.get_url(url % (main_currency + curr))
+                val = res.split(',')[1]
             if val:
                 self.updated_currency[curr] = val
             else:
```

File: currency_rate_update/model/currency_rate_update.py [hunk 2]
```diff
@@ -243,34 +129,24 @@ def refresh_currency(self):
...
-            if not main_currency:
-                # If we can not find a base currency for this company
-                # we look for one with no company set
-                main_currency = curr_obj.search(
-                    [('base', '=', True), ('company_id', '=', False)],
-                    limit=1)
+            main_currency = self.company_id.currency_id
             if not main_currency:
-                raise exceptions.Warning(_('There is no base currency set!'))
+                raise exceptions.Warning(_('There is no main \
```
Comment (yvaucher) [inline currency_rate_update/model/currency_rate_update.py:139]: Please separate the string that way:

```
_('There is no main '
  'currency defined!')
```

Here I think you are introducting many spaces between main and currency

**Comment by yvaucher** [comment]
Please replace  `fields.Datetime.to_string(datetime.today()),` by `fields.Datetime.today()`

https://github.com/OCA/account-financial-tools/pull/363/files#diff-5e8f4cd437749aa1abeff43ff722755aL305

and

https://github.com/OCA/account-financial-tools/pull/363/files#diff-5e8f4cd437749aa1abeff43ff722755aL311

**Comment by yvaucher** [comment]
It seems this doesn't include the last changes https://github.com/OCA/account-financial-tools/pull/318

Ok let's make it as dependent on #318 and merge #318 first

Thanks for the refactoring !

**Comment by yvaucher** [comment]
closing, I'll add this in #318

**Comment by yvaucher** [comment]
@rossigee Could you make a PR for #318 ?

**Pull request closed without merging**